### PR TITLE
CLDR-15034 kbd: new KeyboardTool and KeyboardFlatten utility

### PR DESCRIPTION
**Pull request by srl295**
- 'flattens' keyboards to resolve all <import> statements, with err checking
- correct a bad import in mt.xml
- update docs on implied keys
- add `<key … stretch="true">` 

CLDR-15034

- [ ] This PR completes the ticket.

-----
## Example

`mt.xml` in this repo resolves to:

```xml
<?xml version="1.0" encoding="UTF-8" standalone="no"?><keyboard conformsTo="techpreview" locale="mt">
    <locales>
        <!-- English is also an official language in Malta.-->
        <locale id="en"/>
    </locales>
    <info author="Steven R. Loomis" indicator="MT" layout="QWERTY" normalization="NFC"/>

    <names>
        <name value="Maltese (48-key)"/>
        <!-- <name value="MSA 100:2002" /> -->
    </names>

    <keys>
        <!-- imports -->

<!--Begin Imports from techpreview/keys-Zyyy-punctuation.xml-->
    <!-- General Symbols -->
    <key id="apos" to="'"/>
    <key id="bang" to="!"/>
    <key id="close-paren" to=")"/>
    <key id="double-quote" to="\u{0022}"/>
    <key id="equal" to="="/>
    <key id="hash" to="#"/>
    <key id="hyphen" to="-"/>
    <key id="open-paren" to="("/>
    <key id="umlaut" to="¨"/>
    <key id="underscore" to="_"/>
    <key id="at" to="@"/>
    <key id="backslash" to="\u{005C}"/>
    <key id="caret" to="^"/>
    <key id="close-angle" to="&gt;"/>
    <key id="close-curly" to="}"/>
    <key id="close-square" to="]"/>
    <key id="colon" to=":"/>
    <key id="grave" to="`"/>
    <key id="not" to="¬"/>
    <key id="open-angle" to="\u{003C}"/>
    <key id="open-curly" to="{"/>
    <key id="open-square" to="["/>
    <key id="pipe" to="|"/>
    <key id="plus" to="+"/>
    <key id="question" to="?"/>
    <key id="semi-colon" to=";"/>
    <key id="tilde" to="~"/>
    <key id="amp" to="\u{0026}"/>

    <!-- Currency -->
    <key id="euro" to="€"/>
    <key id="pound" to="£"/>
<!--End Imports from techpreview/keys-Zyyy-punctuation.xml-->

        <!-- accent grave -->
        <key id="a-grave" to="à"/>
        <key id="A-grave" to="À"/>
        <key id="e-grave" to="è"/>
        <key id="E-grave" to="È"/>
        <key id="i-grave" to="ì"/>
        <key id="I-grave" to="Ì"/>
        <key id="o-grave" to="ò"/>
        <key id="O-grave" to="Ò"/>
        <key id="u-grave" to="ù"/>
        <key id="U-grave" to="Ù"/>

        <!-- tikka and maqtua -->
        <key id="c-tikka" to="ċ"/>
        <key id="C-tikka" to="Ċ"/>
        <key id="g-tikka" to="ġ"/>
        <key id="G-tikka" to="Ġ"/>
        <key id="h-maqtua" to="ħ"/>
        <key id="H-maqtua" to="Ħ"/>
        <key id="z-tikka" to="ż"/>
        <key id="Z-tikka" to="Ż"/>

        <!-- Cedilla -->
        <key id="c-cedilla" to="ç"/>

        <!-- gaps -->
        <key gap="true" id="gap" width="1"/>
    </keys>

    <layers form="hardware" hardware="iso">
        <layer modifier="none">
            <row keys="c-tikka 1 2 3 4 5 6 7 8 9 0 minus equals"/>
            <row keys="q w e r t y u i o p g-tikka h-maqtua"/>
            <row keys="a s d f g h j k l semi-colon hash"/>
            <row keys="z-tikka z x c v b n m comma period slash"/>
            <row keys="space"/>
        </layer>

        <layer modifier="shift">
            <row keys="C-tikka bang double-quote euro dollar percent caret amp open-paren close-paren underscore plus"/>
            <row keys="Q W E R T Y U I O P G-tikka H-maqtua"/>
            <row keys="A S D F G H J K L colon at tilde"/>
            <row keys="Z-tikka Z X C V B N M open-angle close-angle question"/>
            <row keys="space"/>
        </layer>

        <layer modifier="altR">
            <row keys="grave gap gap pound gap gap gap gap gap gap gap gap gap"/>
            <row keys="gap gap e-grave gap gap gap u-grave i-grave o-grave gap open-square close-square"/>
            <row keys="a-grave gap gap gap gap gap gap gap gap gap gap gap"/>
            <row keys="backslash gap gap gap gap gap gap gap gap gap gap"/>
            <row keys="space"/>
        </layer>

        <layer modifier="altR-shift">
            <row keys="not gap gap gap gap gap gap gap gap gap gap gap gap"/>
            <row keys="gap gap E-grave gap gap gap U-grave I-grave O-grave gap open-curly close-curly"/>
            <row keys="A-grave gap gap gap gap gap gap gap gap gap gap gap"/>
            <row keys="pipe gap gap gap gap gap gap gap gap gap gap"/>
            <row keys="space"/>
        </layer>
    </layers>
</keyboard>
```

ALLOW_MANY_COMMITS=true
ALLOW_MULTI_COMMITS=true

### CODE REVIEW

**Comment by srl295** [comment]
So an error here is that [NMTOKEN](https://www.w3.org/TR/2000/WD-xml-2e-20000814#NT-BaseChar) is not just ASCII and not just Latin.

File: docs/ldml/tr35-keyboards.md [hunk 1]
```diff
@@ -777,24 +783,40 @@ _Attribute:_ `width="1.2"` (optional, default "1.0")
...
 ```xml
-<!--  all 26 upper and lower case English letters -->
-<key id="a" to="a"/>
-<key id="b" to="b"/>
-<key id="c" to="c"/>
+<key id="gap" gap="true" width="1"/>
+<key id="space" stretch="true" width="1"/>
+```
+
+In addition, these 62 keys where the `id` is the same as the `to` are assumed to exist:
```
Comment (mcdurdin) [inline docs/ldml/tr35-keyboards.md:793]:
```suggestion
In addition, these 62 keys, comprising 10 digit keys, 26 Latin lower-case keys, and 26 Latin upper-case keys, where the `id` is the same as the `to`, are assumed to exist:
```

File: docs/ldml/tr35-keyboards.md [hunk 2]
```diff
@@ -777,24 +783,40 @@ _Attribute:_ `width="1.2"` (optional, default "1.0")
...
-Not all keys need to be listed explicitly.  The following keys can be assumed to already exist:
+Not all keys need to be listed explicitly.  The following two can be assumed to already exist:
 
 ```xml
-<!--  all 26 upper and lower case English letters -->
-<key id="a" to="a"/>
-<key id="b" to="b"/>
-<key id="c" to="c"/>
+<key id="gap" gap="true" width="1"/>
+<key id="space" stretch="true" width="1"/>
```
Comment (mcdurdin) [inline docs/ldml/tr35-keyboards.md:790]:
```suggestion
<key id="space" to=" " stretch="true" width="1"/>
```

We should be a little careful with this one. Some keyboards override spacebar, e.g. Khmer keyboards currently emit ZWSP or COENG MARKER, not space, by default. We may need a paragraph:

> It is permissible to redefine the implied keys.

File: keyboards/import/keys-Zyyy-punctuation.xml [hunk 3]
```diff
@@ -2,36 +2,41 @@
...
+    <key id="percent" to="%" />
+    <key id="period" to="." />
     <key id="pipe" to="|" />
     <key id="plus" to="+" />
     <key id="question" to="?" />
     <key id="semi-colon" to=";" />
+    <key id="slash" to="/" />
     <key id="tilde" to="~" />
-    <key id="amp" to="\u{0026}" />
+    <key id="umlaut" to="¨" />
```
Comment (mcdurdin) [inline keyboards/import/keys-Zyyy-punctuation.xml:35]: I wonder about adding `umlaut` and `not` to this standard file -- I understand that they are on some European keyboards but they are outside the ascii range that all the other punctuation fits into, and may consistute a 'slippery slope' for adding more and more characters to this file? (Same could be argued for `pound` and even `euro` (these could be in a `keys-Zyyy-currency.xml` instead?).

File: tools/cldr-code/src/main/java/org/unicode/cldr/tool/KeyboardTool.java [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+        out.println("<!--\n"
+            + CldrUtility.getCopyrightString() + "\n"
+            + "-->");
+        out.println("<!-- GENERATED by KeyboardTool. This is the implied import for implied keys. -->\n" +
+            "<!DOCTYPE keys SYSTEM \"../dtd/ldmlKeyboard.dtd\">\n" +
+            "<keys>\n");
+
+        out.println("    <key id=\"gap\" gap=\"true\" width=\"1\"/>");
+        out.println("    <key id=\"space\" to=\" \" stretch=\"true\" width=\"1\"/>");
+        out.println();
```
Comment (mcdurdin) [inline tools/cldr-code/src/main/java/org/unicode/cldr/tool/KeyboardTool.java:63]: I'm not a huge fan of xml-by-printf. Wouldn't we be better off with a static, hand written file?

File: tools/cldr-code/src/test/resources/org/unicode/cldr/tool/KeyboardFlatten/broken-import-wrongparent.xml [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+    <!DOCTYPE keyboard SYSTEM "../../../../../../../../../../keyboards/dtd/ldmlKeyboard.dtd"> -->
+<keyboard locale="mt" conformsTo="techpreview">
+    <locales>
+        <!-- English is also an official language in Malta.-->
+        <locale id="en" />
+    </locales>
+    <info author="Steven R. Loomis" normalization="NFC" layout="QWERTY" indicator="MT" />
+
+    <names>
+        <!-- Error - can't import 'keys' into a 'nanes' section. -->
```
Comment (mcdurdin) [inline tools/cldr-code/src/test/resources/org/unicode/cldr/tool/KeyboardFlatten/broken-import-wrongparent.xml:14]:
```suggestion
        <!-- Error - can't import 'keys' into a 'names' section. -->
```

File: tools/cldr-code/src/main/java/org/unicode/cldr/tool/KeyboardTool.java (hunk 4 above)
Comment (srl295) [reply tools/cldr-code/src/main/java/org/unicode/cldr/tool/KeyboardTool.java:63]: perhaps instead we can just retire the file writer and keep the static file.

**Comment by srl295** [comment]
PTAL all. Would like to get this merged so I can fix some other test and data issues.  I know that azerty won't compile for example.

**Comment by srl295** [comment]
rebased yet again

**Review by mcdurdin — APPROVED** [review]

**Pull request merged**